feat(sqlmesh): chaoss metrics and relevant tests

kenyiu opened the pull request:
This is to fix #216  and #4225 

- added metrics `burstiness` https://chaoss.community/kb/metric-burstiness/
- added metrics `contributor_absence_factor` https://chaoss.community/kb/metric-contributor-absence-factor/
- added metrics `prs_comments` https://chaoss.community/kb/metric-change-request-reviews/, which is prerequisites for "project velocity"
- added metrics `project_velocity` https://chaoss.community/kb/metric-project-velocity/
- added metrics `self_merge_rates` https://chaoss.community/kb/metric-self-merge-rates/
- added metrics `new_contributors` https://chaoss.community/kb/metric-new-contributors/
- added metrics `issue_age_avg`, `issue_age_max`, `issue_age_median`, `issue_age_min` https://chaoss.community/kb/metric-issue-age/
- test for `int_events_aux_issues` and fix typos for `int_events_aux_issues`
- use payload.*.updated_at as event_time for several stg models and add tests for them
- tests added for metrics/models touched/related to the changes in this PR.
- hotfix for #4414

ccerv1 commented on warehouse/oso_sqlmesh/models/intermediate/events/github/issues/int_events_aux_issues.sql:
```diff
@@ -115,7 +115,7 @@ WITH github_comments AS (
   FROM oso.stg_github__pull_request_merge_events
   WHERE
-    created_at BETWEEN @start_dt AND @end_dt
+    event_time BETWEEN @start_dt AND @end_dt
```
Note: we should review the comparisons you've done on these two approaches (and the API result)

kenyiu replied:
yes

ccerv1 reviewed (changes requested):
Can you resolve the conflicts in the `metrics_factories.py` ?

ravenac95 reviewed (approved):
@kenyiu looks great! I or @ccerv1 will merge once @ccerv1 is ready. We have some other things we need to push ahead of this since your changes require a fairly large restatement.

Jabolol commented on warehouse/oso_sqlmesh/oso_metrics/code/burstiness.sql:
```diff
@@ -0,0 +1,42 @@
+-- Calculate Fano Factor (variance^2/mean) as a measure of burstiness for repository activity
+-- Manual handling of zero-filled dates using formulas: Average = SUM/COUNT, VAR_POP = SUM(amount^2-mean^2)/COUNT
```
Trino supports `var_pop`, it would be better to use the builtin rather than own impl:

https://trino.io/docs/current/functions/aggregate.html#var_pop

kenyiu replied:
in the first approach, i used `var_pop` and `avg`. the reason why I didnt use them anymore is because the total we want to be divided by is not the total number of rows since there are some days with zero counts

Jabolol commented on warehouse/oso_sqlmesh/oso_metrics/code/burstiness.sql:
```diff
@@ -0,0 +1,42 @@
+    @metrics_sample_date(events.bucket_day) as metrics_sample_date,
+    events.event_source,
+    events.to_artifact_id,
+    sum(events.amount) over (
```
The are two separate window functions with same partition, can you combine them into a single window function?

kenyiu replied:
sure, good catch.

Jabolol commented on warehouse/oso_sqlmesh/oso_metrics/code/burstiness.sql:
```diff
@@ -0,0 +1,42 @@
+    events.bucket_day between @metrics_start('DATE') and @metrics_end('DATE')
+),
+variance_calc as (
+  select
```
You should use the built-in functions: `avg` and `var_pop`. The current impl. is susceptible to floating-point prec. errors, the formula `(Σx² - (Σx)²/n)/n` can produce neg. values due to rounding and the built-in functions are stable

kenyiu replied:
ahhh, i used the wrong formula, no wonder there are negative results. VAR_POP = SUM(amount-mean)^2/COUNT instead of SUM(amount^2-mean^2)/COUNT. i will update it.

Jabolol commented on warehouse/oso_sqlmesh/oso_metrics/code/burstiness.sql:
```diff
@@ -0,0 +1,42 @@
+  @metric_name() as metric,
+  cast(case 
+    when mean_activity > 0 then variance_activity * variance_activity / mean_activity
+    else -1
```
`-1` could mask legitimate zero variance, Consider using NULL or 0

kenyiu replied:
lets do null. there is chances variance_activity = 0

Jabolol commented on warehouse/oso_sqlmesh/oso_metrics/code/project_velocity.sql:
```diff
@@ -0,0 +1,135 @@
+-- Calculate Project Velocity as a weighted combination of development activity metrics
```
The `all_dates_entities` CTE creates a cartesian product by using multiple unions, and adding a distinct, which may result in a huuuge intermediate set. It's generating every possible combo of dates, entities, and sources, which gets crazy expensive. Then it does 4 separate left joins with the same join condition, which makes it tough to optimize or add proper indexes. 

A way better approach would be to just combine all the metrics into one CTE using union all with a metric_type flag, then use a single group by with some conditional aggregation to get all the counts in one go. This would kill the expensive distinct operation, turn four joins into one group by, and perform way better.

Jabolol reviewed (changes requested):
Left some feedback, overall looks good!  Feel free to request another review when they're addressed or ask for questions and I'll gladly go over the feedback with you :)

kenyiu commented on warehouse/oso_sqlmesh/models/metrics_factories.py:
```diff
@@ -707,5 +798,100 @@ def no_gaps_audit_factory(config: MetricQueryConfig) -> tuple[str, dict] | None:
+                "daily",
+                "weekly",
+                "monthly",
+                "quarterly",
```
for those commented out, they are because they are depending on models that have `quarterly` and `biannually` commented out. I assume there's a reason for them to be done that way, so I didnt drill down the reasons but follow.

kenyiu commented:
> Left some feedback, overall looks good! Feel free to request another review when they're addressed or ask for questions and I'll gladly go over the feedback with you :)

@Jabolol clarified and addressed the comments. lemme know your thoughts. thanks 👍

Jabolol reviewed (approved):
LGTM! Thanks for the updates :)
